### PR TITLE
feat: v0 VLM support + GRPO pipeline

### PR DESCRIPTION
**Pull request by rohitrango**
# What does this PR do ?

Adds VLM support (Qwen2.5-VL) with TP plan, DTensor Policy, vLLM backend, and multiple gpus.

# Usage

```python
uv run examples/run_vlm_grpo.py cluster.gpus_per_node=4
```

## Convergence

(Training) convergence on 2 H100 GPUs happens in about 60 iterations. (highest possible reward is 5)

image
image

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](/NVIDIA/NeMo-RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA/NeMo-RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA/NeMo-RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

### CODE REVIEW

**Comment by rohitrango** [comment]
@terrykong @ashors1  Created a draft PR (duplicate of https://github.com/NVIDIA-NeMo/RL/pull/521) to see if CI passes on this instead.

**Comment by terrykong** [comment]
adding @ashors1 @yfw to review/approve

File: examples/run_vlm_grpo.py [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+    # add this for backward compatibility
+    user_message['token_ids'] = message['input_ids'][0]
+    # add all keys and values to the user message, and the list of keys
+    user_message['vlm_keys'] = []
+    for key, value in message.items():
+        # ignore keys (already specified in token_ids)
+        if key in ['input_ids', 'attention_mask']:
+            continue
+        # ignore the batch index if provided
+        user_message[key] = value[0] if key in ['input_ids', 'attention_mask', 'image_grid_thw'] else value
```
Comment (ashors1) [inline examples/run_vlm_grpo.py]: Key will never be in `['input_ids', 'attention_mask']` since we hit the if statement directly above, right?

File: nemo_rl/data/llm_message_utils.py [hunk 2]
```diff
@@ -334,8 +337,24 @@ def batched_message_log_to_flat_message(
...
+        # keys_to_concat = ['pixel_values']
+
+        ## TODO: @rohitrango assuming for now that that all images are the same size (so stacking them is fine)
+        ## this will have consequences for data sharding for VLM models (split along the batch dim but from [start_patch:end_patch])
+        keys_to_concat = []
+
+        if key in keys_to_concat:
+            result[key] = torch.cat(padded)
+        else:
+            result[key] = torch.stack(padded)
```
Comment (ashors1) [inline nemo_rl/data/llm_message_utils.py]: Can this be reverted ?

File: nemo_rl/data/llm_message_utils.py [hunk 3]
```diff
@@ -334,8 +337,24 @@ def batched_message_log_to_flat_message(
...
+            padded = [_pad_tensor(t, max_len, "right", pad_value) for t in filled_values]
+
+        # special case for `pixel_values` which is a flattened (N, D) tensor with potentially different Ns for each message
+        # keys_to_concat = ['pixel_values']
+
+        ## TODO: @rohitrango assuming for now that that all images are the same size (so stacking them is fine)
+        ## this will have consequences for data sharding for VLM models (split along the batch dim but from [start_patch:end_patch])
+        keys_to_concat = []
+
+        if key in keys_to_concat:
```
Comment (yfw) [inline nemo_rl/data/llm_message_utils.py]: keys_to_concat is always empty here?

File: nemo_rl/distributed/batched_data_dict.py [hunk 4]
```diff
@@ -512,7 +511,19 @@ def repeat_interleave(self, num_repeats: int) -> Self:
...
+        if len(vlm_keys) > 0:
+            # this is a dict of vlm_keys to their tensors, so we need to collect all the keys
+            if isinstance(vlm_keys[0], dict):
+                vlm_keys = [key for subdict in vlm_keys for key in subdict.keys()]
+                vlm_keys = list(set(vlm_keys))
+
         for k, v in self.items():
+            # skip truncation for VLM keys
+            if k in vlm_keys:
+                self.data[k] = v
```
Comment (ashors1) [inline nemo_rl/distributed/batched_data_dict.py]: Is this necessary?

File: nemo_rl/data/hf_datasets/clevr.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    image.save(buffered, format=format)
+    img_str = base64.b64encode(buffered.getvalue()).decode("utf-8")
+    # append data:image/png;base64,
+    return f"data:image/png;base64,{img_str}"
+
+def format_answer_fromtags(answer: str) -> str:
+    """
+    Format the answer from tags.
+    """
+    # Extract content between <answer> tags and strip whitespace
```
Comment (yfw) [inline nemo_rl/data/hf_datasets/clevr.py]: I think this is a better docstring than the current one

File: nemo_rl/data/hf_datasets/refcoco.py [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+
+    # resize image for easy image processing across batches
+    image = Image.open(example['image_path'])
+    orig_width, orig_height = image.size
+    resized_image = image.resize((width, height))
+    flip = False
+
+    if random.random() < img_flip_prob:
+        flip = True
+        resized_image = resized_image.transpose(Image.FLIP_LEFT_RIGHT)
```
Comment (yfw) [inline nemo_rl/data/hf_datasets/refcoco.py]: Is this always safe to do for this dataset? Do any captions rely on the positions of the original image (e.g. "A cat sitting to the left of a dog")

File: nemo_rl/models/generation/vllm.py [hunk 7]
```diff
@@ -600,13 +619,30 @@ async def generate_async(
...
+            else:
+                # Regular LLM generation using token_ids
+                # Prepare prompt token IDs for this specific sample
+                current_input_actual_length = input_lengths_batch[sample_idx].item()
+                prompt_token_ids_list = (
+                    input_ids_batch[sample_idx, :current_input_actual_length].tolist()
+                    if current_input_actual_length > 0
+                    else []
+                )
+                prompt = {"prompt_token_ids": prompt_token_ids_list}
```
Comment (ashors1) [inline nemo_rl/models/generation/vllm.py]: Looks like this code block is replicated for `generate` and `generate_async`. Can we refactor to avoid duplicated code?

File: nemo_rl/models/generation/vllm_backend.py [hunk 8]
```diff
@@ -106,6 +106,13 @@ def update_weights_from_ipc_handles(self, ipc_handles):
             print(
                 f"Error in VllmInternalWorkerExtension.update_weights_from_ipc_handles: {e}"
             )
+            # print the stack trace
+            import traceback
+
+            print("Model runner:")
+            print(self.model_runner.model)
+
+            traceback.print_exc()
```
Comment (ashors1) [inline nemo_rl/models/generation/vllm_backend.py]: Remove debug code?

File: nemo_rl/models/policy/dtensor_policy_worker.py [hunk 9]
```diff
@@ -434,7 +457,14 @@ def train(
         # dim 1 is always assumed to be the sequence dim, sanity check this here
         sequence_dim = 1
         seq_dim_size = data.get("input_ids").shape[sequence_dim]
+
+        # get vlm keys from data
+        vlm_keys = get_vlm_keys_from_clippedpgloss_batch(data)
```
Comment (ashors1) [inline nemo_rl/models/policy/dtensor_policy_worker.py]: Can `get_vlm_keys_from_clippedpgloss_batch` be renamed to reflect the fact that it's independent of the particular loss function? e.g. `get_vlm_keys_from_clippedpgloss_batch` is called if we're running SFT, too

File: nemo_rl/models/policy/dtensor_policy_worker.py [hunk 10]
```diff
@@ -894,9 +938,8 @@ def prepare_weights_for_ipc(self) -> tuple[list[tuple[str, int]], float]:
 
         # Get state_dict
         self.model = self.move_to_cuda(self.model)
-        self._held_sharded_state_dict_reference: dict[str, torch.Tensor] = (
-            self.model.state_dict()
-        )
+
+        self._held_sharded_state_dict_reference = self.model.state_dict()
```
Comment (ashors1) [inline nemo_rl/models/policy/dtensor_policy_worker.py]: revert this?

File: examples/run_vlm_grpo.py [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+        add_generation_prompt=True,
+    )
+    if isinstance(string_formatted_dialog, list):
+        string_formatted_dialog = string_formatted_dialog[0]
+
+    # add this for backward compatibility
+    user_message['token_ids'] = message['input_ids'][0]
+    # add all keys and values to the user message, and the list of keys
+    user_message['vlm_keys'] = []
+    for key, value in message.items():
```
Comment (yfw) [inline examples/run_vlm_grpo.py]: Are the vlm_keys specific to the dataset? And are they applicable for all messages of that dataset? If so, can this be configured with the dataset? (i.e. in clevr.py and refcoco.py). This seems to assume all keys except for 'input_ids', 'attention_mask' are vlm_keys which seems less safe than if we were explicit about which keys are vlm keys.

**Comment by yfw** [comment]
I tried a run on the clevr dataset and noticed the token_mult_prob_error was a bit high. This is a measure of the difference in logprobs between vllm and dtensor. For non-vlm, we generally expect this number to be < 1.05 (for qwen 2.5-1b, we see around 1.02). A higher number usually indicates some issue with the refit so vllm and dtensor aren't exactly matching. Is there anything specific about the vlm setup that could cause this?

Screenshot 2025-07-18 at 6 30 17 PM

**Comment by rohitrango** [comment]
I'm not entirely sure what causes the `token_mult_prob_error ` to be relatively higher than the LLM only case. I did not fiddle with the either vllm or the dtensor model forward setup so I'm not sure what causes this.

For the functional test cases, I had to choose higher thresholds for the `token_mult_prob_error` errors. There are also a few spikes like you have shown that I cannot quite explain (the loss at these iterations does not spike though).

File: nemo_rl/models/dtensor/parallelize.py [hunk 12]
```diff
@@ -41,6 +41,18 @@
 from transformers.models.qwen2.modeling_qwen2 import Qwen2ForCausalLM
 from transformers.models.qwen3.modeling_qwen3 import Qwen3ForCausalLM
 
+# Add VL model imports
+try:
+    from transformers.models.qwen2_vl.modeling_qwen2_vl import Qwen2VLForConditionalGeneration, Qwen2_5_VLModel
```
Comment (yfw) [inline nemo_rl/models/dtensor/parallelize.py]: Do we need these try/excepts? Can we instead make sure the transformers version we're using has these?

File: examples/run_vlm_grpo.py (hunk 11 above)
Comment (yfw) [reply examples/run_vlm_grpo.py]: Can we get the whitelist of vlm_keys from `processor.image_processor.model_input_names` ?

**Comment by rohitrango** [comment]
This will take me a while to analyse since I don't know exactly how the vllm engine processes the images internally.

For the policy, the typical multimodal pipeline is to use the `processor` to encode the chat template dict into a sequence of text tokens, multimodal tokens indexed by `pixel_values` and metadata `image_grid_thw` to compute mRoPE embeddings (keys would be different for videos or audio).

For vllm, the message log is simply reformatted into the format specified in [this tutorial](https://docs.vllm.ai/en/v0.9.1/features/multimodal_inputs.html#image-inputs). I assume the token ids and multimodal tokens must be the same, but I will have to double check. In most cases the logprobs mult error is still close to 1, so it could also be numerical differences between the vllm frontend and the dtensor policy.

**Comment by rohitrango** [comment]
re: Remaining blockers:

1. *understanding the logprob error*: This is something I want to chalk up to how vllm loads multimodal image embeddings in the image processor. For LLM-only, I noted that vllm takes the same list of `token_ids` (int value list) that the policy consumes (i.e. going through the same text embedding layer, etc.). However, for multimodal images, vllm processes the images internally. There could also be differences in how sampling is done differently. I found the following excerpt from vllm docs https://docs.vllm.ai/en/v0.9.1/usage/v1_guide.html#feature-model

> Logprobs in V1 are now returned immediately once computed from the model’s raw output (i.e. before applying any logits post-processing such as temperature scaling or penalty adjustments). As a result, the returned logprobs do not reflect the final adjusted probabilities used during sampling.
> Support for logprobs with post-sampling adjustments is in progress and will be added in future updates.

I prefer handling this issue in a separate PR (and merging an initial support) for three reasons:

* this discrepancy is isolated to multimodal models only, so a "fix" can be shipped independently
* multiple VLMs converge on three different datasets _despite_ the apparent discrepancy. It is equivalent to training GRPO with a slightly off-policy model, but it does not seem to be very unstable or destructive to the learning process
* the PR has gotten very big as it is, and adding more fixes will add additional overhead to the review process
* other PRs break multimodal support regularly (every 2-3 days) and I have to rollback those changes to make my scripts work. Merging this PR at least the test cases will prevent further breaking changes

2. PR has now migrated (again) to https://github.com/NVIDIA-NeMo/RL/pull/712 . #712 is tested on 4 families of multimodal models and 3 datasets. This rollsback the `vlm_kwargs` list business and instead proposes a `PackedGenericDataItem` to handle non-sequence data items (most of them would be multimodal tensors). The single implementation seems to work for multiple multimodal models without any additional modifications to the config.
